src/chains/ethereum/console.log/scripts/helpers.ts

```diff
@@ -0,0 +1,322 @@
+import { readFileSync } from "fs";
+import { join } from "path";
+
+const CONTRACT_NAME = "ConsoleLogTest";
+const consoleSol = readFileSync(join(__dirname, "../", "console.sol"), "utf8");
+
+export type SignatureDetail = {
+  solidity?: string;
+  javascript?: string;
+  params: string[];
+  name: string;
+};
+
+export type FunctionDescriptor = {
+  functionName: string;
+  params: { type: string }[];
+  consoleSolLogFunctionToCall: string;
+};
+
+export const combinatorTypes = [
+  "address",
+  "bool",
+  "string memory",
+  "uint256"
+] as const;
+
+// for compatibility with hardhat's console.log, which uses `int` instead of
+// `int256`, we need to also include `int` aliases in the permutations.
+type SolidityType = typeof primitiveTypes[number] | BytesN | "uint" | "int";
+
+const primitiveTypes = [...combinatorTypes, "bytes memory", "int256"] as const;
+
+/**
+ * Hardhat abi encodes uint instead of uint256. This saves a couple of bytes,
+ * but is incorrect.
+ */
+export const hardhatTypeAliases: Map<SolidityType, SolidityType> = new Map([
+  ["uint256", "uint"],
+  ["int256", "int"]
+]);
+
+const typeToHandlerMap: Map<SolidityType, Handler> = new Map([
+  ["address", "address"],
+  ["bool", "bool"],
+  ["bytes memory", "bytes"],
+  ["int", "int256"],
+  ["int256", "int256"],
+  ["string memory", "string"],
+  ["uint", "uint256"],
+  ["uint256", "uint256"],
+  // generate bytes1 .. bytes32
+  ...(Array.from({ length: 32 }, (_, i) => [
+    `bytes${i + 1}` as BytesN,
+    `fixedBytes(${i + 1})` as FixedBytesN
+  ]) as [BytesN, FixedBytesN][])
+]);
+
+const COMMENT = `
+    /**
+    * Prints to \`stdout\` with newline. Multiple arguments can be passed, with the
+    * first used as the primary message and all additional used as substitution
+    * values similar to [\`printf(3)\`](http://man7.org/linux/man-pages/man3/printf.3.html) (the arguments are all passed to \`util.format()\`).
+    *
+    * \`\`\`solidity
+    * uint256 count = 5;
+    * console.log('count: %d', count);
+    * // Prints: count: 5, to stdout
+    * console.log('count:', count);
+    * // Prints: count: 5, to stdout
+    * \`\`\`
+    *
+    * See \`util.format()\` for more information.
+    */`;
+
+/**
+ * Cache used to ensure we do not accidentally generate 4 byte collisions.
+ */
+const signatureCache: Map<number, string> = new Map();
+
+/**
+ * Generates the solidity and javascript function signatures for a given set of
+ * params.
+ * @param params
+ * @param solidityFunctionName
+ * @returns
+ */
+function getSignature(
+  params: SolidityType[],
+  solidityFunctionName = "log"
+): SignatureDetail {
+  const abiParams = params.map(type => type.replace(" memory", ""));
+  const abiSignatureString = `log(${abiParams.join(",")})`;
+  // the solidity "4-bytes" signature:
+  const signature = keccak(Buffer.from(abiSignatureString)).subarray(0, 4);
+
+  // we store the signature as an int on the JS side:
+  const signatureInt = parseInt(signature.toString("hex"), 16);
```

Comment:
buffer.toString accepts a `start` and `end` param, so instead of creating a subarray (granted this is cheap, and might be more clear), we could go direct with:

`keccak(Buffer.from(abiSignatureString)).toString("hex", 0, 4)`

Reply:
I do _not_ feel strongly about this.